src/types.ts

```diff
@@ -94,55 +91,54 @@ export interface PublicConfiguration<
 
 export type FullConfiguration = InternalConfiguration & PublicConfiguration
 
-export type ConfigOptions = {
+export type ProviderConfiguration = {
   initFocus: (callback: () => void) => (() => void) | void
   initReconnect: (callback: () => void) => (() => void) | void
 }
+
 export interface SWRHook {
-  <Data = any, Error = any, Args extends Key = Key>(args: Args): SWRResponse<
+  <Data = any, Error = any, SWRKey extends Key = Key>(key: SWRKey): SWRResponse<
     Data,
     Error
   >
-  <Data = any, Error = any, Args extends Key = Key>(
-    args: Args,
-    fn: Fetcher<Data, Args> | null
+  <Data = any, Error = any, SWRKey extends Key = Key>(
+    key: SWRKey,
+    fetcher: Fetcher<Data, SWRKey> | null
   ): SWRResponse<Data, Error>
-  <Data = any, Error = any, Args extends Key = Key>(
-    args: Args,
-    config: SWRConfiguration<Data, Error, Args, Fetcher<Data, Args>> | undefined
+  <Data = any, Error = any, SWRKey extends Key = Key>(
+    key: SWRKey,
+    config: SWRConfiguration<Data, Error, SWRKey> | undefined
   ): SWRResponse<Data, Error>
-  <Data = any, Error = any, Args extends Key = Key>(
-    args: Args,
-    fn: Fetcher<Data, Args>,
-    config: SWRConfiguration<Data, Error, Args, Fetcher<Data, Args>>
+  <Data = any, Error = any, SWRKey extends Key = Key>(
+    key: SWRKey,
+    fetcher: Fetcher<Data, SWRKey>,
+    config: SWRConfiguration<Data, Error, SWRKey>
   ): SWRResponse<Data, Error>
-  <Data = any, Error = any, Args extends Key = Key>(
+  <Data = any, Error = any, SWRKey extends Key = Key>(
     ...args:
-      | [Args]
-      | [Args, Fetcher<Data, Args> | null]
-      | [
-          Args,
-          SWRConfiguration<Data, Error, Args, Fetcher<Data, Args>> | undefined
-        ]
+      | [SWRKey]
+      | [SWRKey, Fetcher<Data, SWRKey> | null]
+      | [SWRKey, SWRConfiguration<Data, Error, SWRKey> | undefined]
       | [
-          Args,
+          SWRKey,
           Fetcher<Data, Key> | null,
-          SWRConfiguration<Data, Error, Args, Fetcher<Data, Args>>
+          SWRConfiguration<Data, Error, SWRKey>
         ]
   ): SWRResponse<Data, Error>
 }
 
 // Middlewares guarantee that a SWRHook receives a key, fetcher, and config as the argument
-type SWRHookWithMiddleware = <Data = any, Error = any, Args extends Key = Key>(
+export type Middleware = (
+  useSWRNext: SWRHook
+) => <Data = any, Error = any, Args extends Key = Key>(
   key: Args,
   fetcher: Fetcher<Data, Args> | null,
   config: SWRConfiguration<Data, Error>
 ) => SWRResponse<Data, Error>
 
-export type Middleware = (useSWRNext: SWRHook) => SWRHookWithMiddleware
-export type TupleKey = [any, ...unknown[]] | readonly [any, ...unknown[]]
-export type ValueKey = string | null | TupleKey | Record<any, any>
-export type Key = ValueKey | (() => ValueKey)
+type ArgumentsTuple = [any, ...unknown[]] | readonly [any, ...unknown[]]
```

Comment:
small Q: is readonly here necessary?

Reply:
Yeah I think they're different for TS tuples, cc @promer94 who initially implemented it.

Reply:
Great catch ! It could be removed.
```typescript
// the origin implementation looked like this
type Fetcher<Args extends Key> = Args extends (() => [...infer K] | readonly [...infer K])
  ? ((...args: [...K]) => Result<Data>) : ...

// But then shu made me realize the it could have batter implementation
// So the Args does not have to satisfy [any, ...unknown[]] | readonly [any, ...unknown[]] any more
type Fetcher<Args extends Key> = Args extends (() => readonly [...infer K] | null)
  ? ((...args: [...K]) => Result<Data>) : ...
```